### PR TITLE
Do not bind if there is no binding item collected from the service

### PR DESCRIPTION
**Pull request by pmacik**
### Motivation

Issue #1030

### Changes

Do not bind as files if there is no binding data is collected from the service

### CODE REVIEW

File: pkg/reconcile/pipeline/handler/project/impl.go [hunk 1]
```diff
@@ -70,6 +70,11 @@ func BindingsAsEnv(ctx pipeline.Context) {
 		return
 	}
 
+	if len(ctx.BindingItems()) == 0 {
```
Comment (pedjak) [inline pkg/reconcile/pipeline/handler/project/impl.go:73]: we should put this check in `PreFlightCheck` function.

File: pkg/reconcile/pipeline/handler/project/impl.go [hunk 2]
```diff
@@ -70,6 +70,11 @@ func BindingsAsEnv(ctx pipeline.Context) {
...
+		stop(ctx, errors.New("no binding item to project"))
```
Comment (pedjak) [inline pkg/reconcile/pipeline/handler/project/impl.go:74]: `stop` means that we are not going to retry, but we should actually.

File: pkg/reconcile/pipeline/handler/project/impl.go [hunk 3]
```diff
@@ -160,6 +165,10 @@ func BindingsAsFiles(ctx pipeline.Context) {
 	if !ctx.BindAsFiles() {
 		return
 	}
+	if len(ctx.BindingItems()) == 0 {
+		stop(ctx, errors.New("no binding item to project"))
+		return
+	}
```
Comment (pedjak) [inline pkg/reconcile/pipeline/handler/project/impl.go:171]: see above comment on `PreFlightCheck` - once there, this check can be removed.

File: test/acceptance/features/bindAppToService.feature [hunk 4]
```diff
@@ -842,3 +842,30 @@ Feature: Bind an application to a service
...
+            metadata:
+                name: $scenario_id
+            spec:
+                bindAsFiles: false
+                detectBindingResources: true
+                services:
+                  - group: ""
+                    version: v1
+                    kind: Service
+                    name: $scenario_id
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:862]: we did not create service earlier to bind to?

File: test/acceptance/features/injectBindingsAsFiles.feature [hunk 5]
```diff
@@ -275,6 +275,33 @@ Feature: Bindings get injected as files in application
...
+            metadata:
+                name: $scenario_id
+            spec:
+                bindAsFiles: true
+                detectBindingResources: true
+                services:
+                  - group: ""
+                    version: v1
+                    kind: Service
+                    name: $scenario_id
```
Comment (pedjak) [inline test/acceptance/features/injectBindingsAsFiles.feature:294]: same as above

**Review by pedjak — CHANGES_REQUESTED** [review]

File: test/acceptance/features/bindAppToService.feature (hunk 4 above)
Comment (pmacik) [reply test/acceptance/features/bindAppToService.feature:862]: The service is created along with the generic test application, but for clarity I can add a separate service so it is clear in the scenario that the service is present

File: pkg/reconcile/pipeline/handler/project/impl.go (hunk 1 above)
Comment (pmacik) [reply pkg/reconcile/pipeline/handler/project/impl.go]: +1

File: pkg/reconcile/pipeline/handler/project/impl.go (hunk 3 above)
Comment (pmacik) [reply pkg/reconcile/pipeline/handler/project/impl.go]: +1

File: test/acceptance/features/bindAppToService.feature (hunk 4 above)
Comment (pmacik) [reply test/acceptance/features/bindAppToService.feature]: Done

File: test/acceptance/features/injectBindingsAsFiles.feature (hunk 5 above)
Comment (pmacik) [reply test/acceptance/features/injectBindingsAsFiles.feature]: Done, thx

File: pkg/reconcile/pipeline/handler/project/impl.go (hunk 2 above)
Comment (pmacik) [reply pkg/reconcile/pipeline/handler/project/impl.go]: good point, thanks

File: pkg/reconcile/pipeline/handler/project/impl.go [hunk 6]
```diff
@@ -40,6 +40,13 @@ func PreFlightCheck(mandatoryBindingKeys ...string) func(pipeline.Context) {
 					return
 				}
 			}
+		} else {
+			if len(items) == 0 {
+				err := errors.New("no binding item to project")
+				ctx.RetryProcessing(err)
+				ctx.SetCondition(apis.Conditions().NotInjectionReady().Reason(apis.NoBindingItemsReason).Msg(err.Error()).Build())
+				return
```
Comment (pedjak) [inline pkg/reconcile/pipeline/handler/project/impl.go:48]: go conventions try to avoid `else` blocks. do this check right after line 31 in a separate if then block.

File: test/acceptance/features/bindAppToService.feature [hunk 7]
```diff
@@ -570,6 +570,8 @@ Feature: Bind an application to a service
             kind: Backend
             metadata:
                 name: demo-backserv-cr-3
+                annotations:
+                    service.binding/name: path={.metadata.name}
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:574]: how is this change related to PR change?

File: test/acceptance/features/bindAppToService.feature [hunk 8]
```diff
@@ -842,3 +848,43 @@ Feature: Bind an application to a service
...
+                    name: $scenario_id-svc
+                application:
+                    name: $scenario_id
+                    group: apps
+                    version: v1
+                    resource: deployments
+            """
+        Then Service Binding CollectionReady.status is "True"
+        * Service Binding InjectionReady.status is "False"
+        * Service Binding InjectionReady.message is "no binding item to project"
```
Comment (pedjak) [inline test/acceptance/features/bindAppToService.feature:890]: I would perhaps change the message to `no binding data to project`

**Review by pedjak — CHANGES_REQUESTED** [review]

File: test/acceptance/features/bindAppToService.feature (hunk 7 above)
Comment (pmacik) [reply test/acceptance/features/bindAppToService.feature:574]: The PR makes it necessary to have some binding data in order to finish binding successfully and to make it 'Ready'. Those 3 scenarios did not have anything bindable and so with introducing this PR it failed at first SB being checked for Readiness. So I have to introduce some binding data for those scenarios to work.
